### PR TITLE
Avoid creating transaction inside transaction (compressed streams)

### LINKED ISSUE #1451: loadStreamAsFile creates Russian Doll Transaction for compressed streams
Related to #1443 - the compressing stream store solution #1357 creates a transaction inside a transaction when loading a stream and dumping it directly to a file. This is because it reuses `makeStream`, which has the BlockGetter as a performance optimisation.

Lazily loading the stream has no benefit when we're immediately loading it anyway, so this should be refactored.

### PR DESCRIPTION
**Pull request by gsheasby**
Enforce in makeStream's BlockGetter that the original transaction is committed

This ensures that in StreamStore, we don't create a transaction inside a transaction.

This exposes an issue with compressedStream, specifically when it eagerly loads the stream, so that it can save the decompressed stream to a file. This is also fixed here.

Fixes #1451 

FAO @sandorw.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1460)

### CODE REVIEW

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/generated/StreamTestWithHashStreamStore.java [hunk 1]
```diff
@@ -214,13 +214,22 @@ public InputStream loadStream(Transaction t, final Long id) {
...
     protected void tryWriteStreamToFile(Transaction transaction, Long id, StreamMetadata metadata, FileOutputStream fos) throws IOException {
-        try (InputStream blockStream = makeStream(id, metadata);
+        try (InputStream blockStream = eagerlyLoadStream(transaction, id, metadata);
                 InputStream decompressingStream = new LZ4BlockInputStream(blockStream);
                 OutputStream fileStream = fos;) {
             ByteStreams.copy(decompressingStream, fileStream);
         }
     }
 
+    private InputStream eagerlyLoadStream(Transaction parent, Long id, StreamMetadata metadata) {
```
Comment (sandorw) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/generated/StreamTestWithHashStreamStore.java]: The nice thing about the previous implementation (using makeStream) is that we actually stream from cassandra to the file. This implementation holds the entire stream contents (albeit compressed) in memory.

Comment (gsheasby) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/generated/StreamTestWithHashStreamStore.java]: However, it did this with an inner transaction, which wasn't allowed. I guess an alternative would be to reuse the `parent` transaction in the compressing case.

Comment (sandorw) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/generated/StreamTestWithHashStreamStore.java]: Yep, that makes sense - I'm just trying to make sure we avoid holding the entire stream in memory if we can.

Comment (gsheasby) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/generated/StreamTestWithHashStreamStore.java]: Sure. I'll have a look at implementing it that way tomorrow.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/StreamStoreRenderer.java [hunk 2]
```diff
@@ -571,14 +575,41 @@ private void loadStreamsWithCompression() {
...
                     } line("}");
                 } line("}");
             }
 
+            private void makeStreamUsingTransaction() {
+                line("private InputStream makeStreamUsingTransaction(Transaction parent, ", StreamId, " id, StreamMetadata metadata) {"); {
+                    line("long totalBlocks = getNumberOfBlocksFromMetadata(metadata);");
+                    line("int blocksInMemory = getNumberOfBlocksThatFitInMemory();");
+                    line();
+                    line("BlockGetter pageRefresher = new BlockGetter() {"); {
```
Comment (SerialVelocity) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/StreamStoreRenderer.java]: Move this into a separate file so only one class has to be loaded by java.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java [hunk 3]
```diff
@@ -100,17 +101,20 @@ private InputStream tryGetStream(Transaction transaction, ID id, StreamMetadata
...
 
-    protected InputStream makeStream(ID id, StreamMetadata metadata) {
+    private InputStream makeStream(Transaction parent, ID id, StreamMetadata metadata) {
         long totalBlocks = getNumberOfBlocksFromMetadata(metadata);
         int blocksInMemory = getNumberOfBlocksThatFitInMemory();
 
         BlockGetter pageRefresher = new BlockGetter() {
             @Override
             public void get(long firstBlock, long numBlocks, OutputStream destination) {
+                Preconditions.checkState(!parent.isUncommitted(),
```
Comment (SerialVelocity) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: Test to make sure this is hit?

**Pull request merged**